File: 144_binaryTreePreorderTraveral/Solution.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
    	std::vector<int> v;
    	return this->auxPreorderTraversal(root, v);
    }

    vector<int>& auxPreorderTraversal(TreeNode* root, vector<int>& res) {
    	if (root == NULL) return res;
        res.push_back(root->val);
    	if (root->left != NULL) {
    		this->auxPreorderTraversal(root->left, res);
    	}
    	if (root->right != NULL) {
    		this->auxPreorderTraversal(root->right, res);
    	}
    	return res;
    }
};

File: 144_BinaryTreePreorderTraveral/Solution.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def __init__(self):
        self.res = []
    def preorderTraversal(self, root):
        if not root: return self.res
        self.res.append(root.val)
        if root.left:
            self.preorderTraversal(root.left)
        if root.right:
            self.preorderTraversal(root.right)
        return self.res
